sw/device/tests/crypto/cryptotest/firmware/ibex_fi.c

```diff
@@ -0,0 +1,103 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/tests/crypto/cryptotest/firmware/ibex_fi.h"
+
+#include "sw/device/lib/base/memory.h"
+#include "sw/device/lib/base/status.h"
+#include "sw/device/lib/runtime/log.h"
+#include "sw/device/lib/testing/test_framework/ottf_test_config.h"
+#include "sw/device/lib/testing/test_framework/ujson_ottf.h"
+#include "sw/device/lib/ujson/ujson.h"
+#include "sw/device/sca/lib/sca.h"
+#include "sw/device/tests/crypto/cryptotest/json/ibex_fi_commands.h"
+
+#define NOP1 "addi x0, x0, 0\n"
+#define NOP10 NOP1 NOP1 NOP1 NOP1 NOP1 NOP1 NOP1 NOP1 NOP1 NOP1
+#define NOP100 NOP10 NOP10 NOP10 NOP10 NOP10 NOP10 NOP10 NOP10 NOP10 NOP10
+
+#define INITX5 "addi x5, x0, 0"
+
+#define ADDI1 "addi x5, x5, 1\n"
+#define ADDI10 ADDI1 ADDI1 ADDI1 ADDI1 ADDI1 ADDI1 ADDI1 ADDI1 ADDI1 ADDI1
+#define ADDI100 \
+  ADDI10 ADDI10 ADDI10 ADDI10 ADDI10 ADDI10 ADDI10 ADDI10 ADDI10 ADDI10
+#define ADDI1000                                                          \
+  ADDI100 ADDI100 ADDI100 ADDI100 ADDI100 ADDI100 ADDI100 ADDI100 ADDI100 \
+      ADDI100
+
+/**
+ * Initializes the SCA trigger.
+ *
+ *
+ * @param uj The received uJSON data.
+ */
+status_t handle_ibex_fi_init_trigger(ujson_t *uj) {
+  // TODO: For now, use the AES SCA trigger. In the future, we want to
+  // introduce a FI specific trigger.
+  sca_select_trigger_type(kScaTriggerTypeSw);
+  sca_init(kScaTriggerSourceAes, kScaPeripheralIoDiv4 | kScaPeripheralAes);
+  return OK_STATUS(0);
+}
+
+/**
+ * ibex.char.unrolled_reg_op_loop command handler.
+ *
+ * This FI penetration tests executes the following instructions:
+ * - Initialize register x5=0
+ * - Add 100 NOPs to delay the trigger
+ * - Perform 10000 x5 = x5 + 1 additions
+ * - Return the value over UART.
+ *
+ * Faults are injected during the trigger_high & trigger_low.
+ * It needs to be ensured that the compiles does not optimize this code.
+ *
+ * @param uj The received uJSON data.
+ */
+status_t handle_ibex_fi_char_unrolled_reg_op_loop(ujson_t *uj) {
+  uint32_t loop_counter = 0;
+  sca_set_trigger_high();
+  asm volatile(INITX5);
+  asm volatile(NOP100);
+  asm volatile(ADDI1000);
+  asm volatile(ADDI1000);
+  asm volatile(ADDI1000);
+  asm volatile(ADDI1000);
+  asm volatile(ADDI1000);
+  asm volatile(ADDI1000);
+  asm volatile(ADDI1000);
+  asm volatile(ADDI1000);
+  asm volatile(ADDI1000);
+  asm volatile(ADDI1000);
+  asm volatile("mv %0, x5" : "=r"(loop_counter));
+  sca_set_trigger_low();
+
+  // Send loop counter to host.
+  ibex_fi_loop_counter_t uj_output;
+  uj_output.loop_counter = loop_counter;
+  RESP_OK(ujson_serialize_ibex_fi_loop_counter_t, uj, &uj_output);
+  return OK_STATUS(0);
+}
```

Comment:
Somewhere around here, we could/should read some relevant status registers to see if an alert has been raised. There could be a test-specific section, in this case [rv_core_ibex.ERR_STATUS](https://opentitan.org/book/hw/ip/rv_core_ibex/doc/registers.html#err_status), as well as a general section reading from the alert handler. The general should probably be a library call and I assume it's a bit more work to add that. So we don't need to do this in this very PR. But the specific one would be nice to have in this first example.